src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows.Controls;
+using Microsoft.CodeAnalysis.Editor.Implementation.Adornments;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
+{
+    internal class InlineDiagnosticsAdornmentManager : AbstractAdornmentManager<InlineDiagnosticsTag>
+    {
+        private readonly IClassificationTypeRegistryService _classificationRegistryService;
+        private readonly IClassificationFormatMap _formatMap;
+
+        public InlineDiagnosticsAdornmentManager(
+            IThreadingContext threadingContext, IWpfTextView textView, IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IAsynchronousOperationListener asyncListener, string adornmentLayerName,
+            IClassificationFormatMapService classificationFormatMapService,
+            IClassificationTypeRegistryService classificationTypeRegistryService)
+            : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
+        {
+            _classificationRegistryService = classificationTypeRegistryService;
+            _formatMap = classificationFormatMapService.GetClassificationFormatMap(textView);
+            _formatMap.ClassificationFormatMappingChanged += OnClassificationFormatMappingChanged;
+            TextView.ViewportWidthChanged += TextView_ViewportWidthChanged;
+        }
+
+        /// <summary>
+        /// Need to remove the tags if they intersect with the editor view, but only if the option
+        /// to place the tags at the end of the editor is selected.
+        /// </summary>
+        private void TextView_ViewportWidthChanged(object sender, EventArgs e)
+        {
```

Comment:
can you do an access check that you're on the UI thread?

Reply:
done